rfcs/0008-storage-handler.md

```diff
@@ -0,0 +1,338 @@
+- Feature Name: Storage Handlers
+- Start Date: 2023-02-10)
+- RFC PR: [FuelLabs/sway-rfcs#0023](https://github.com/FuelLabs/sway-rfcs/pull/023)
+- Sway Issue: [FueLabs/sway#3043](https://github.com/FuelLabs/sway/issues/3043)
+
+# Summary
+
+[summary]: #summary
+
+This RFC introduces the concept of a `StorageHandler` which describes a particular storage slot via its key. The RFC also reworks how storage accesses are represented in the language and how `StorageHandler` can be used to build dynamic storage types, such as `StorageMap` and `StorageVec`, more robustly.
+
+# Motivation
+
+[motivation]: #motivation
+
+The current approach for reading and writing storage slots/variables has multiple flaws, despite being quite ergonomic:
+
+1. There is currently no way to reason about the location of a particular storage variable, which makes passing "references" to storage variables impossible. One workaround is manipulating and reading storage slots manually via `std::storage::store` and `std::storage::get`, but this approach is quite unsafe. Alternatively, one could declare a `trait` in a library and implement it in a contract and have its methods access storage variables. Those methods are then used in the library to access storage indirectly. This does not solve the full problem and feels more like a workaround than a real solution.
+1. Dynamic storage types, such as `StorageMap` and `StorageVec`, currently require a "hacky" compiler intrinsic called `__get_storage_key` which is not very well defined and is hard to use. Introducing the concept of a `StorageHandler` will helps us completely remove this intrinsic.
+1. The current implementation of dynamic storage types prevents th-em from being used as struct fields or as type parameters of other storage types (e.g. `StorageVec<StorageVec<u64>>`).
```

Comment:
3.

Reply:
All 3 points above are labeled `1.` :)

Reply:
That's how Markdown works. :)  If you look at the rendered version they are sequential.  Makes it easier to edit the list, to insert new items later on.

Reply:
To Toby's point. I see that we do this in the Fuel specs so I followed the same approach here.